agent/src/capability/servicemanifestloader.h

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2017 Pelagicore AB
+ *
+ * Permission to use, copy, modify, and/or distribute this software for
+ * any purpose with or without fee is hereby granted, provided that the
+ * above copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+ * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR
+ * BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES
+ * OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
+ * WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
+ * ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
+ * SOFTWARE.
+ *
+ * For further information see LICENSE
+ */
+
+
+#pragma once
+
+#include "softwarecontainer-common.h"
+
+#include <jsonparser.h>
+
+namespace softwarecontainer {
+
+/*
+ * Abstract interface to a service manifest loader.
+ *
+ * The loader will either be a production loader that loads a json file
+ * from a file on the system, or a test loader that loads json from a string.
+ */
+class ServiceManifestLoader
+{
+public:
+    // Enforce inheriting classes to initialize service manifest source member
+    ServiceManifestLoader() = delete;
+
+    /*
+     * @brief Constructs a service manifest loader object and stores the source string
+     * @param string could either be a string representing e.g. the json or
+     * the path to the service manifest(s) depending on the inheriting class
+     */
+    ServiceManifestLoader(const std::string &source) : m_source(source) {}
+
+    /*
+     * The destructor decref:s the json items in the content so that all
+     */
+    virtual ~ServiceManifestLoader() {
+        for (json_t *item : m_content) {
+            json_decref(item);
+        }
+    }
+
+    /*
+     * @brief Loads the json content of the service manifest(s) based on the source
+     * sent to the constructor.
+     *
+     * @return a vector of json_t corresponding to the content of the service manifest
+     * @throws ConfigStoreError depending on the implementation of the inheriting
+     * classes
```

Comment:
I don't understand. This is the interface that tells the user what will happen and it tells anyone inheriting this class what happens so they could implement it accordingly. An interface should not allow the implementing class to choose, it should follow what the interface dictates.